### PR TITLE
sql/schemachanger: implement DROP OWNED BY

### PR DESCRIPTION
**Pull request by jasonmchan**
Previously, we did not support the DROP OWNED BY statement (#55381).
This commit adds partial support for DROP OWNED BY in the declarative
schema changer. Followup work is needed to support the CASCADE modifier.

Release note (sql change): Support `DROP OWNED BY`.

### CODE REVIEW

**Review by Xiang-Gu — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82936)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jasonmchan and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4cj-NG04AYNLhowDjk:-N4cj-NG04AYNLhowDjl:b-hkyg9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L44)):*
> ```Go
> 	var toCheckBackrefs []descpb.ID
> 
> 	// Lookup all objects in the current database.
> ```

curious: can a user own objects across databases? If so, what's the expected behavior if we `DRO OWNED BY him`?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4ckcpQ0TNeYQwB4NVm:-N4ckcpQ0TNeYQwB4NVn:bad52vi) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L56)):*
> ```Go
> 		})
> 		objects = append(objects, sp.SchemaID)
> 	})
> ```

Is this block used to retrieve all object IDs in `db`?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4clVzt6zC6ZeuRleqK:-N4clVzt6zC6ZeuRleqL:b-k65ar9) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L61)):*
> ```Go
> 	for _, id := range objects {
> 		elts := b.QueryByID(id)
> 		_, _, owner := scpb.FindOwner(elts)
> ```

is it possible that `elts` does not have a Owner element?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4ciech5630dco80ckR:-N4ciech5630dco80ckS:b-8cwxv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L88)):*
> ```Go
> 		for _, role := range normalizedRoles {
> 			if e.UserName == role.Normalized() && e.UserName != b.SessionData().User().Normalized() {
> 				dropElement(b, e)
> ```

you probably need to change this to simply `b.Drop(e)` with #82907 coming out soon.
___
*[`pkg/sql/logictest/testdata/logic_test/drop_owned_by` line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4chDyXDvZPry5aMAQO:-N4chDyY-a-rPO_IG8cx:b-ai5qi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/logictest/testdata/logic_test/drop_owned_by#L83)):*
> ```
> SHOW TABLES FROM public
> ----
> ```

1. when `testuser` drops all objects he owned, why are `u` and `v1` also dropped (which are owned by `testuser2`）?

2. in addition to dropping the objects, we also revoke the privileges from users. Do we tests this behavior in the logic test?
___
*[`pkg/sql/logictest/testdata/logic_test/drop_owned_by` line 212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4chw0J0VXlw78v5zOo:-N4chw0J0VXlw78v5zOp:b-i5ipql) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/logictest/testdata/logic_test/drop_owned_by#L212)):*
> ```
> 
> statement error pq: cannot drop objects owned by role "admin" because they are required by the database system
> DROP OWNED BY admin
> ```

Can we add some more comments on the test? It's a bit hard to follow and remember what's been dropped and what's been created by which user. 

You can break up the tests into subtests and each subtest focus on a pattern/theme/case of tests.

**Review by Xiang-Gu — COMMENTED** [review]
Good job! Keep up the good work! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-:-N4cmknX3v7qEZ6hc4fJ:b-93rhl0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jasonmchan and @Xiang-Gu)

**Review by jasonmchan — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-:-N4haTXg8vmU4BETKZ-Z:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4cj-NG04AYNLhowDjk:-N4hZtF9A5aZ3AP4Xbxo:b5uk2ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L44)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

curious: can a user own objects across databases? If so, what's the expected behavior if we `DRO OWNED BY him`?
</blockquote></details>

Yes. DROP OWNED BY drops objects in the current database (see [Postgres docs](https://www.postgresql.org/docs/current/sql-drop-owned.html))
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4ckcpQ0TNeYQwB4NVm:-N4h_CgNEy0RBNsRRZEa:btne1vo) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L56)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Is this block used to retrieve all object IDs in `db`?
</blockquote></details>

Yes, would it be clearer if I renamed `objects` to `objectIDs`?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4clVzt6zC6ZeuRleqK:-N4h_PGc1apD4hMAmZ8o:buk1bm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L61)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

is it possible that `elts` does not have a Owner element?
</blockquote></details>

I've been working under this assumption, but maybe someone else can confirm. My understanding is that [all objects have owners](https://www.cockroachlabs.com/docs/v22.1/security-reference/authorization#object-ownership), which is enforced when we [decompose descriptors into elements](https://github.com/cockroachdb/cockroach/blob/1f6c94c57ca2878f15e459faf0f832299832ebf1/pkg/sql/schemachanger/scdecomp/decomp.go#L83-L86).
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4ciech5630dco80ckR:-N4h_dGPA5MU0mRDlDYM:b-b0isqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L88)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

you probably need to change this to simply `b.Drop(e)` with #82907 coming out soon.
</blockquote></details>

Thanks for letting me know! I didn't pay close attention to the difference :sweat_smile: 
___
*[`pkg/sql/logictest/testdata/logic_test/drop_owned_by` line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4chDyXDvZPry5aMAQO:-N4h_koQDzpfI1azMyrF:b-dnvvk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/logictest/testdata/logic_test/drop_owned_by#L83)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

1. when `testuser` drops all objects he owned, why are `u` and `v1` also dropped (which are owned by `testuser2`）?

2. in addition to dropping the objects, we also revoke the privileges from users. Do we tests this behavior in the logic test?
</blockquote></details>

1. The previous statement `DROP OWNED BY testuser2` drops them. The intention here was to test `CASCADE`; I've added a TODO comment to indicate this.
2. Yes
___
*[`pkg/sql/logictest/testdata/logic_test/drop_owned_by` line 212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4chw0J0VXlw78v5zOo:-N4haJDd1kGPgmAXBdh8:b-l61z8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/logictest/testdata/logic_test/drop_owned_by#L212)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Can we add some more comments on the test? It's a bit hard to follow and remember what's been dropped and what's been created by which user. 

You can break up the tests into subtests and each subtest focus on a pattern/theme/case of tests.
</blockquote></details>

I've attempted to improve the readability of the tests. Hopefully they're clearer now!

**Review by chengxiong-ruan — COMMENTED** [review]
Nice work :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-:-N55lChK3MC1iTyAFUFH:b90cghs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @jasonmchan, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/dependencies.go` line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55hpjR5Bw6convgTA9:-N55hpjS6YHLOZiPyRZv:be0u3tr) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/dependencies.go#L172)):*
> ```Go
> 
> 	// MemberOfWithAdminOption looks up the roles that 'member' is a member of.
> 	MemberOfWithAdminOption(ctx context.Context, member username.SQLUsername) (map[username.SQLUsername]bool, error)
> ```

the go convention is to use `map[username.SQLUsername]struct{}` here to avoid the confusion of `true` and `false` boolean values :)
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go` line 177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55iC1JA2QfKFvOUC3e:-N55iC1K2I4c61l_QkDV:bohcslz) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go#L177)):*
> ```Go
> 	// CheckMemberOf returns true iff the current user has membership in the
> 	// specified role.
> 	IsMemberOf(member username.SQLUsername) bool
> ```

`IsMemberOf` -> `CurrentUserIsMemberOf`?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N4ciech5630dco80ckR:-N55kFds5kNYkfL6_gGT:b3xuxan) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb377e65415847c643f023c2fa597744c256e188/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L88)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

Thanks for letting me know! I didn't pay close attention to the difference :sweat_smile: 
</blockquote></details>

ha, nice
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55j8EWEZKvDLJ7wtTr:-N55j8EWEZKvDLJ7wtTs:b-cudtut) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L27)):*
> ```Go
> // DropOwnedBy implements DROP OWNED BY.
> func DropOwnedBy(b BuildCtx, n *tree.DropOwnedBy) {
> 	normalizedRoles, err := decodeusername.FromRoleSpecList(
> ```

what does `normalized` mean here?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55jJxHA97tHvsgdr2J:-N55jJxHA97tHvsgdr2K:b-6c9co2) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L36)):*
> ```Go
> 		if role.IsAdminRole() || role.IsRootUser() || role.IsNodeUser() {
> 			panic(pgerror.Newf(pgcode.DependentObjectsStillExist,
> 				"cannot drop objects owned by role %q because they are required by the database system", role))
> ```

Is it legit to drop non-system objects though?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55jZAK7CcgNH7IlYo8:-N55jZAK7CcgNH7IlYo9:b-i7o5ls) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L39)):*
> ```Go
> 		}
> 		if role != b.SessionData().User() && !b.IsMemberOf(role) {
> 			panic(pgerror.New(pgcode.InsufficientPrivilege, "permission denied to drop objects"))
> ```

Is it legit for admin/root perform this operation if they are not part the input `roles`?
___
*[`pkg/sql/logictest/testdata/logic_test/schema_change_feature_flags` line 249 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55ke3O1aC5nOps4RTj:-N55ke3O1aC5nOps4RTk:bcaihji) ([raw file](https://github.com/cockroachdb/cockroach/blob/aadbaf97b4e6092ad6978a28e2735715d64d9f10/pkg/sql/logictest/testdata/logic_test/schema_change_feature_flags#L249)):*
> ```
> REASSIGN OWNED BY root TO testuser
> 
> # Test DROP OWNED BY.
> ```

Could you add `DROP OWNED BY` a special case here as well so that it still use its `SchemaFeatureName` which is `DROP OWNED BY`?: https://github.com/cockroachdb/cockroach/blob/4ee236cc83d14f7aea41c82afab3ecc9c0f00095/pkg/sql/sem/tree/schema_feature_name.go#L26-L26 ?
We'd need to refactor this better somehow. I haven't thought about it yet :(

**Review by jasonmchan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82936)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/dependencies.go` line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55hpjR5Bw6convgTA9:-N57FMiH5j7P_OfQU93A:b5x32u4) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/dependencies.go#L172)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

the go convention is to use `map[username.SQLUsername]struct{}` here to avoid the confusion of `true` and `false` boolean values :)
</blockquote></details>

My comment was misleading, the boolean value actually indicates whether the member is an admin of the role. I've updated the comment. Leaving the signature as is to maintain parity with the `AuthorizationAccessor` in `authorization.go`.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go` line 177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55iC1JA2QfKFvOUC3e:-N57FKri9zJPB-xDf9CI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go#L177)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

`IsMemberOf` -> `CurrentUserIsMemberOf`?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55j8EWEZKvDLJ7wtTr:-N57FDQB18W58NFEafoo:bh6pcey) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L27)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

what does `normalized` mean here?
</blockquote></details>

It's referring to the [username being normalized](https://github.com/cockroachdb/cockroach/blob/75cf2bb581fdc57db0928d0e907d94003e760c04/pkg/security/username/username.go#L24-L60).
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55jJxHA97tHvsgdr2J:-N57BhfRDu521o_c39lj:b-szropl) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L36)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Is it legit to drop non-system objects though?
</blockquote></details>

Postgres won't drop non-system objects for these roles. I'm not sure if there's a case where these roles don't own any system objects and so the statement would succeed, but I think that erroring here is the safe option.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55jZAK7CcgNH7IlYo8:-N57CtwwBYrcE6CxdqWB:b-3mwn2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L39)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Is it legit for admin/root perform this operation if they are not part the input `roles`?
</blockquote></details>

Yes. The check for this is in the `CurrentUserIsMemberOf` function, although maybe it would be clearer if the check were elsewhere. Here's a [logictest](https://github.com/cockroachdb/cockroach/blob/f93df508d8c3d2fe97e722ea3774874d11dd614b/pkg/sql/logictest/testdata/logic_test/drop_owned_by#L179-L182) for this scenario.
___
*[`pkg/sql/logictest/testdata/logic_test/schema_change_feature_flags` line 249 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55ke3O1aC5nOps4RTj:-N577LxO9ARenwS-W9x4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aadbaf97b4e6092ad6978a28e2735715d64d9f10/pkg/sql/logictest/testdata/logic_test/schema_change_feature_flags#L249)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Could you add `DROP OWNED BY` a special case here as well so that it still use its `SchemaFeatureName` which is `DROP OWNED BY`?: https://github.com/cockroachdb/cockroach/blob/4ee236cc83d14f7aea41c82afab3ecc9c0f00095/pkg/sql/sem/tree/schema_feature_name.go#L26-L26 ?
We'd need to refactor this better somehow. I haven't thought about it yet :(
</blockquote></details>

Done.

**Review by chengxiong-ruan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82936)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @jasonmchan, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/dependencies.go` line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55hpjR5Bw6convgTA9:-N5AhFvQE2O_G8CNFEXO:b9p87p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/dependencies.go#L172)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

My comment was misleading, the boolean value actually indicates whether the member is an admin of the role. I've updated the comment. Leaving the signature as is to maintain parity with the `AuthorizationAccessor` in `authorization.go`.
</blockquote></details>

ah, interesting. thanks for explaining that.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55jJxHA97tHvsgdr2J:-N5AiKPu23CVkV3c7WuV:b-ik71dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L36)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

Postgres won't drop non-system objects for these roles. I'm not sure if there's a case where these roles don't own any system objects and so the statement would succeed, but I think that erroring here is the safe option.
</blockquote></details>

yeah, that make sense, thanks.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55jZAK7CcgNH7IlYo8:-N5Aktnx0uvsmsgV3qEX:b-bdvxi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L39)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

Yes. The check for this is in the `CurrentUserIsMemberOf` function, although maybe it would be clearer if the check were elsewhere. Here's a [logictest](https://github.com/cockroachdb/cockroach/blob/f93df508d8c3d2fe97e722ea3774874d11dd614b/pkg/sql/logictest/testdata/logic_test/drop_owned_by#L179-L182) for this scenario.
</blockquote></details>

Hmm. wait.. I thought `CurrentUserIsMemberOf` only checks if current is a member of `role`, but don't really care about if it's an admin or not?
Oh.. does `MemberOfWithAdminOption` actually also return `admin` even the `role` is not an admin? heh...
___
*[`pkg/sql/schemachanger/scplan/testdata/drop_owned_by` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N5B21H36tjByFFLYCLr:-N5B21H36tjByFFLYCLs:b-o275ot) ([raw file](https://github.com/cockroachdb/cockroach/blob/44f23264fa8f9ce7e36bd183298c5f2605359c4e/pkg/sql/schemachanger/scplan/testdata/drop_owned_by#L1)):*
> ```
> setup
> ```

Sorry, I should have noticed this earlier but just realized that we need a builder test [similar to these](https://github.com/cockroachdb/cockroach/tree/master/pkg/sql/schemachanger/scbuild/testdata) as well

**Review by jasonmchan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82936)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @jasonmchan, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55jZAK7CcgNH7IlYo8:-N5GJHQ83wqMu-2JRRe7:b-wqs1c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L39)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Hmm. wait.. I thought `CurrentUserIsMemberOf` only checks if current is a member of `role`, but don't really care about if it's an admin or not?
Oh.. does `MemberOfWithAdminOption` actually also return `admin` even the `role` is not an admin? heh...
</blockquote></details>

It's actually because I put a separate check [here](https://github.com/cockroachdb/cockroach/blob/3dac5168551702c740a389b452c183e9bbb567c3/pkg/sql/schemachanger/scbuild/builder_state.go#L161) for the admin case. I renamed the function to `CurrentUserHasAdminOrIsMemberOf`, is this better?
___
*[`pkg/sql/schemachanger/scplan/testdata/drop_owned_by` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N5B21H36tjByFFLYCLr:-N5GJem1-nif8SYZcwjB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44f23264fa8f9ce7e36bd183298c5f2605359c4e/pkg/sql/schemachanger/scplan/testdata/drop_owned_by#L1)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Sorry, I should have noticed this earlier but just realized that we need a builder test [similar to these](https://github.com/cockroachdb/cockroach/tree/master/pkg/sql/schemachanger/scbuild/testdata) as well 
</blockquote></details>

Done.

**Review by chengxiong-ruan — APPROVED** [review]
:lgtm_strong: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-:-N5GaXJfAdVXEDw7Pnn8:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @jasonmchan, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82936#-N55jZAK7CcgNH7IlYo8:-N5Ga3E20fB2uzutnuBC:b2nqqzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/37304b7b51fc81e4c4026990d58b1ae619799d0b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L39)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

It's actually because I put a separate check [here](https://github.com/cockroachdb/cockroach/blob/3dac5168551702c740a389b452c183e9bbb567c3/pkg/sql/schemachanger/scbuild/builder_state.go#L161) for the admin case. I renamed the function to `CurrentUserHasAdminOrIsMemberOf`, is this better?
</blockquote></details>

Thanks that's much better!

File: pkg/sql/schemachanger/scop/mutation.go [hunk 1]
```diff
@@ -530,6 +530,13 @@ type RemoveDatabaseRoleSettings struct {
 	DatabaseID descpb.ID
 }
 
+// RemoveUserPrivileges is used to revoke a user's privileges.
+type RemoveUserPrivileges struct {
+	mutationOp
+	DescID descpb.ID
+	User   string
+}
```
Comment (postamar) [inline pkg/sql/schemachanger/scop/mutation.go:538]: User privileges are implicitly removed when dropping something, so you can remove this op altogether.

**Review by postamar — APPROVED** [review]
This looks good, nicely done! I have a few comments but nothing big, they should be straightforward for you to address

**Comment by postamar** [comment]
bors r-

Sorry for the late review! Please address my comments and feel free to re-bors.

File: pkg/sql/schemachanger/scop/mutation.go (hunk 1 above)
Comment (jasonmchan) [reply pkg/sql/schemachanger/scop/mutation.go:538]: I believe `DROP OWNED BY` should revoke user privileges for the role(s) on objects that aren't owned, so we need this op to drop those privileges without dropping the entire objects.

Comment (chengxiong-ruan) [reply pkg/sql/schemachanger/scop/mutation.go:538]: Good catch. @jasonmchan see rules in `pkg/sql/schemachanger/scplan/internal/rules/op_drop.go`

Comment (jasonmchan) [reply pkg/sql/schemachanger/scop/mutation.go:538]: I think [this rule](https://github.com/cockroachdb/cockroach/blob/2461c0883b2440997ed48130b183ba371f0490ef/pkg/sql/schemachanger/scplan/internal/rules/op_drop.go#L202-L239) should handle this, since user privileges are part of the descriptor.

Comment (chengxiong-ruan) [reply pkg/sql/schemachanger/scop/mutation.go:538]: Amazing. Good find.

**Review by Xiang-Gu — APPROVED** [review]

File: pkg/sql/schemachanger/scop/mutation.go (hunk 1 above)
Comment (postamar) [reply pkg/sql/schemachanger/scop/mutation.go:538]: Excellent!

File: pkg/sql/logictest/testdata/logic_test/drop_owned_by [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+DROP OWNED BY testuser2
+
+statement error pq: permission denied to drop objects
+DROP OWNED BY testuser, testuser2
+
+statement error pq: cannot drop objects owned by role "root" because they are required by the database system
+DROP OWNED BY root
+
+statement error pq: cannot drop objects owned by role "admin" because they are required by the database system
+DROP OWNED BY admin
```
[the remark this answers is not in the document]
Comment (jasonmchan) [reply pkg/sql/logictest/testdata/logic_test/drop_owned_by:393]: I changed the unimplemented error from the one in `scerror` to the one in `unimplemented`, since it seems like the former causes us to try to fall back to the old schema changer, which ends up throwing a less descriptive error message for `DROP OWNED BY`. I added the logictest check but not the `unimplemented_drop` one--is this the right call?

[the remark this answers is not in the document]
Comment (chengxiong-ruan) [reply pkg/sql/logictest/testdata/logic_test/drop_owned_by:393]: Yeah, I hear you. I'm good with this approach since we're going to implement it anyway. But I'm wondering if you could just set new schema changer's mode to `unsafe_always` since you disable the legacy schema changer configuration in this logic test :) also cc @postamar if you have strong opinion on the options here.

[the remark this answers is not in the document]
Comment (postamar) [reply pkg/sql/logictest/testdata/logic_test/drop_owned_by:393]: Yes, absolutely. Thanks.

**Comment by ajwerner** [comment]
I think this has some merge skew on generated output.

bors r-

**Comment by jasonmchan** [comment]
Merge conflict should be fixed!

bors r+

**Comment by RichardJCai** [comment]
Can we close https://github.com/cockroachdb/cockroach/issues/55381 with this?

We have a separate issue for support CASCADE.
https://github.com/cockroachdb/cockroach/issues/55908

**Comment by postamar** [comment]
By all means!
